Code/Mantid/Framework/Algorithms/src/VesuvioL1ThetaResolution.cpp

```diff
@@ -0,0 +1,458 @@
+
+  m_sample = m_instWorkspace->getInstrument()->getSample();
+}
+
+//----------------------------------------------------------------------------------------------
+/** Loads the instrument into a workspace.
+ */
+void VesuvioL1ThetaResolution::calculateDetector(IDetector_const_sptr detector, std::vector<double>& l1Values, std::vector<double>& thetaValues) {
+  const int numEvents = getProperty("NumEvents");
+  l1Values.reserve(numEvents);
+  thetaValues.reserve(numEvents);
+
+  double sampleWidth = getProperty("SampleWidth");
+  // If the sample is large fix the width to the approximate beam width
+  if(sampleWidth > 4.0)
+    sampleWidth = 4.0;
+
+  // Get detector dimensions
+  Geometry::Object_const_sptr pixelShape = detector->shape();
+  if (!pixelShape || !pixelShape->hasValidShape()) {
+    throw std::invalid_argument("Detector pixel has no defined shape!");
+  }
+  Geometry::BoundingBox detBounds = pixelShape->getBoundingBox();
+  V3D detBoxWidth = detBounds.width();
+  const double detWidth = detBoxWidth.X() * 100;
+  const double detHeight = detBoxWidth.Y() * 100;
+
+  g_log.debug() << "detWidth=" << detWidth << std::endl
+                << "detHeight=" << detHeight << std::endl;
+
+  // Scattering angle in rad
+  const double theta = m_instWorkspace->detectorTwoTheta(IDetector_const_sptr(detector));
+  if(theta == 0.0)
+    return;
+
+  // Final flight path in cm
+  const double l1av = detector->getDistance(*m_sample) * 100.0;
+
+  const double x0 = l1av * sin(theta);
+  const double y0 = l1av * cos(theta);
+
+  // Get as many events as defined by NumEvents
+  // This loop is not iteration limited but highly unlikely to ever become infinate
+  while(l1Values.size() < static_cast<size_t>(numEvents)) {
+    const double xs = -sampleWidth/2 + sampleWidth*random();
+    const double ys = 0.0;
+    const double zs = -sampleWidth/2 + sampleWidth*random();
+    const double rs = sqrt(pow(xs, 2) + pow(zs, 2));
+
+    if(rs <= sampleWidth/2) {
+      const double a = -detWidth/2 + detWidth*random();
+      const double xd = x0 - a*cos(theta);
+      const double yd = y0 + a*sin(theta);
+      const double zd = -detHeight/2 + detHeight*random();
+
+      const double l1 = sqrt(pow(xd-xs, 2) + pow(yd-ys, 2) + pow(zd-zs, 2));
+      double angle = acos(yd / l1);
+
+      if(xd < 0.0)
+        angle *= -1;
+
+      // Convert angle to degrees
+      angle *= 180.0 / M_PI;
+
+      l1Values.push_back(l1);
+      thetaValues.push_back(angle);
+    }
+
+    interruption_point();
+  }
+}
+
+//----------------------------------------------------------------------------------------------
+/** Rebins the distributions and sets error values.
+ */
+MatrixWorkspace_sptr VesuvioL1ThetaResolution::processDistribution(MatrixWorkspace_sptr ws, const double binWidth) {
+  const size_t numHist = ws->getNumberHistograms();
+
+  double xMin(DBL_MAX);
+  double xMax(DBL_MIN);
+  for(size_t i = 0; i < numHist; i++) {
+    const std::vector<double> x = ws->readX(i);
```

Comment:
This would be better off as a `const std::vector<double> &`, so that the X values are not copied.